### PR TITLE
Refactor locations managements

### LINKED ISSUE #4712: Proposal the new way to manage Locations in BHIMA
The way the BHIMA system managed localization was very fixed and was not flexible to manage the different architectures possible to manage localization.
The architecture we had used looks like this

**Country / Province / Sector / Village**
 
this architecture is very limited if you want to save the locations for, for example, the city of Kinshasa which looks like this.
**Country / Province / District / Municipality / District / Street**

If we take the example of existing territorial entities such as for example the city of Kananga where we had already installed the system, we should have the following architecture

**Country / Province / City / Municipality / Village** but during the installation we had forced the system by defining Kananga as being a sector whereas Kananga is a City and the city in the current architecture are at the same level as the territory that we had not managed, e we had registered Tshikaji as a village which is normal but Tshikaji is found in the commune of NGANZA

image

With only the Democratic Republic of Congo the way we manage localization has many limits because we can end up with several types of architecture

• Country / Province / District / Municipality / District
• Country / Province / District / Municipality / District / Street
• Country / Province / District / Municipality / Village
• Country / Province / Territory / City
• Country / Province / Territory / Municipality
• Country / Province / Territory / Municipality / Village
• Country / Province / Territory / Sector / Village

And if we take the example of France for example we will have the following architecture
• Country / Region / Department / Municipality
And if we take the example of the United States of America we will have the following architecture
• Country / State / City / District
By these examples we realize that we can have a standard architecture, reason why we should have a flexible system

Reason for distinguishing the place tree and their respective status, we propose the creation of a table to manage the different types of location and another which will be like a tree structure with the Country entities which will be like the roots. And we should define the elements that we should consider as leaves

Here is an example of the elements that there will be in the location table type

image

Here is an example of an overview of the location table.
To better manage this architecture, it would be necessary to design a flexible and standard component which could respond to this architecture compared to existing configurations.

image

### PR DESCRIPTION
**Pull request by lomamech**
- Refactor the management of Locations
- Add location_type manage API and Component
- Add location management
- Implement component bhLocatioConfigurationSelect
- Implement component bhLocationTypeSelect
- remove modules for country, province, sector and village management
- remove component bhLocationSelect
- Test integration for Location and location_type management

closes #4712

### CODE REVIEW

File: server/controllers/admin/locations.js [hunk 1]
```diff
@@ -16,410 +12,488 @@
...
+        SELECT DISTINCT(t.id) AS id, t.translation_key, t.color, t.label_name
+        FROM location_type AS t
+        JOIN location AS l ON l.location_type_id = t.id
+        ${clauseWhere}
+        ORDER BY l.name ASC;`;
+
+      let sqlHeader = ``;
+      let sqlJoin = ``;
+
+      // Here we try to add indices for each level of localization the indices start from 1
```
[the remark this answers is not in the document]
Comment (jmcameron) [reply server/controllers/admin/locations.js:277]: This is exciting!

**Comment by jmcameron** [comment]
I'm starting to look at this PR.  One comment/suggestion (so far).  On the Location Management Configuration page, I think that the "Merge Locations" button will rarely be used, so I think it would be good to not make it quite as prominent.  For instance, you could create a "Menu" button on the far right and move the "Merge Locations" function into the "Menu" drop-down menu.  That way users only see the "Add" button at first glance.   Typical users are not distracted by the "Merge Locations" button, and people familiar enough with BHIMA to be doing merge operations can find it easily.

**Comment by jmcameron** [comment]
In the Locations Management list of locations, it would be helpful to include indenting of the location names like the "Account Management" table (http://localhost:8080/#!/accounts). Having all the names left-justified does not give a sense of what belongs to what.

**Comment by jmcameron** [comment]
I also like the way that the Accounts Management table includes the "Add Child Acount" after the  account name rather than have a separate column with just a '+'.   If you did the same in the locations manager with a "Add Child Location" that would be very clear to the user.
image

**Comment by jmcameron** [comment]
Sorry, I was referring to the "Location Manager Configuration" page when I was suggesting indenting, etc.

**Comment by jmcameron** [comment]
I found some debugging comments including "NEWWWWww".  These should be cleaned up.

**Comment by jniles** [comment]
@lomamech, I know you put a lot of work into this PR and it represents a huge change.  For the moment, I think all our energy needs to be focused on stock management work, so we'll just have to let this one hang out for a while.  It should be merged after this stock push!  Looks like you have gotten some good feedback so far.

**Comment by lomamech** [comment]
> @lomamech, I know you put a lot of work into this PR and it represents a huge change. For the moment, I think all our energy needs to be focused on stock management work, so we'll just have to let this one hang out for a while. It should be merged after this stock push! Looks like you have gotten some good feedback so far.



I know, that's why I haven't dealt with @jmcameron 's comments so far

**Comment by jmcameron** [comment]
@lomamech I suggest you put this PR into DRAFT status so it does not appear on our current list of open PRs.    When we get to the point we are ready to bring it in, we can re-enable it.

**Comment by jmcameron** [comment]
Postpone this for now.

**Pull request closed without merging**